src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs

```diff
@@ -466,5 +466,147 @@ public void WriteWithSpanEndingAfterEndThrowsOutOfRange()
 
             Assert.Equal("span", ex.ParamName);
         }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71006")]
+        public void Comparer_EqualityContract_SameContent_DifferentEncodings()
+        {
+            // Test case from the issue: two SourceText instances with same content but different encodings
+            // should be equal and have the same hash code when using content-based comparison
+            const string content = "Hello, World!";
+            var utf8 = Encoding.UTF8;
+            var unicode = Encoding.Unicode;
+
+            var text1 = SourceText.From(content, utf8);
+            var text2 = SourceText.From(content, unicode);
+
+            var comparer = SourceTextComparer.Instance;
+
+            // Both should be considered equal by the comparer (content-based)
+            Assert.True(comparer.Equals(text1, text2));
+
+            // They must have the same hash code to satisfy IEqualityComparer contract
+            Assert.Equal(comparer.GetHashCode(text1), comparer.GetHashCode(text2));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71006")]
+        public void Comparer_EqualityContract_SameContent_WithAndWithoutBOM()
+        {
+            // Test the case described in the issue comments:
+            // Create source texts from byte arrays where one has a BOM and the other doesn't
+            const string content = "Test content";
+
+            var utf8WithBOM = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true);
```

Comment:
It looks like this flag only controls the return value of `GetPreamble()`. You could consider deleting the `UTF8Encoding` locals and just using `Encoding.UTF8` in this test.